proposals/0003-external-calls.md

```diff
@@ -72,165 +83,195 @@ HoneyPot :: caller <- (any) {
 
-try! boundReturn <- alpha!.withdraw() then {
-  // Optionally does something with boundReturn
-}
+ - Type states cannot be specified
+ - Caller protection blocks cannot be specified
+ - `mutating` or `public` keywords cannot be specified on functions
+ - Default implementations cannot be specified
+ - `Self` cannot be used
 
-// Setting hyper parameters //
-// Asset types are atomically transferred to preserve special properties.
-alpha!.transfer(Wei(200))
-alpha!.allocate(Gas(2000))
-alpha.trust() // Removes the need for a bang
-try! alpha.expensiveFunction()
-```
+Some additional caveats:
 
-### Trusted Calls
+ - Function arguments can be given labels, but these are for internal use only (since they do not affect the ABI signature)
+ - Functions can be given return types, but there is no trivial way to check if a returned value is of the required type (e.g. a `Bool` `true` value has the same representation as a `Int` `1` in the Solidity ABI)
+ - External traits have an implicit constructor, so that an address can be "cast" into the trait, allowing function calls
+ - Functions of external trait instances cannot be called using the regular function call syntax, but must use the `call` keyword, which also allows hyper-parameters to be specified
 
-#### Flint Package Manager
 ```swift
-// Creates a contract from the data stored in Flint Package Manager
-import flint://0x000... as ERCToken
-
-ERCToken.transfer(...)
+external trait Alpha {
+  func simpleFunction()
+  func functionWithArguments(value: Int, tax: Int)
+  func functionWithReturn() -> Int
+  
+  @payable
+  func expensiveFunction()
+}
 ```
-#### Source Code
-A contract's source code can be imported by:
-- Directly downloading its source files
-- Providing a web URL
-- Finding the Flint contract in the (future) Flint Package Manager
 
-```swift
-import https://flint.org/examples/contract.flint as URLContract
-import FileContract
-import Directory.Contract
+The trait can then be used in Flint code:
 
-let contract: Contract<URLContract> = deploy(URLContract)
-contract.argumentName() // Value and Gas are automatically set based upon properties
+```swift
+contract Beta {
+  func useAlpha() {
+    let someAddress: Address = 0x... // deployed Alpha contract
+    let alpha: Alpha = Alpha(address: someAddress) // using the implicit constructor
+    
+    do {
+      call alpha.simpleFunction() // with default gas allocation: 2300
+      call(gas: 200) alpha.simpleFunction() // with user-specified gas allocation
+    } catch ExternalCallError {
+      // recover gracefully
+    }
+    
+    // if the call fails, this will terminate
+    call! alpha.functionWithArguments(value: 1, tax: 2)
+    
+    // error: user must specify an amount of Wei to pay
+    // call! alpha.expensiveFunction()
+    
+    call(wei: 100)! alpha.expensiveFunction()
+    
+    if let returnedValue: Int = call alpha.functionWithReturn() {
```

Comment:
If we follow swift convention we should have a `call?` construct that returns `nil` if the call fails. In swift, doing 

`if let returnedValue: Int = try bla bla`

Would actually expect an optional to be returned from the function rather than because an exception was (or wasn't) thrown.